sinatra-contrib/lib/sinatra/capture.rb

```diff
@@ -84,40 +84,27 @@ module Sinatra
   module Capture
     include Sinatra::EngineTracking
 
-    DUMMIES = {
-      :haml   => "!= capture_haml(*args, &block)",
-      :erubi  => "<% @capture = yield(*args) %>",
-      :erubis => "<% @capture = yield(*args) %>",
-      :slim   => "== yield(*args)"
-    }
-
     def capture(*args, &block)
-      @capture = nil
-      if current_engine == :ruby
-        result = block[*args]
-      elsif current_engine == :erb || current_engine == :slim
-        @_out_buf, _buf_was = '', @_out_buf
-        raw = block.call(*args)
-        captured = block.binding.eval('@_out_buf')
-        result = captured.empty? ? raw : captured
-        @_out_buf = _buf_was
+      if ruby?
+        block[*args]
+      elsif haml?
+        buffer = Haml::Buffer.new(nil, Haml::Options.new.for_buffer)
+        with_haml_buffer(buffer) { capture_haml(*args, &block) }
       else
-        buffer     = eval '_buf if defined?(_buf)', block.binding
-        old_buffer = buffer.dup if buffer
-        dummy      = DUMMIES.fetch(current_engine)
-        options    = { :layout => false, :locals => {:args => args, :block => block }}
-
-        buffer.clear unless buffer.nil?
-        result = render(current_engine, dummy, options, &block)
+        @_out_buf, _buf_was = '', @_out_buf
+        begin
+          raw = block[*args]
+          captured = block.binding.eval('@_out_buf')
+          captured.empty? ? raw : captured
+        ensure
+          @_out_buf = _buf_was
+        end
       end
-      result && result.strip.empty? && @capture ? @capture : result
-    ensure
-      buffer.replace(old_buffer) unless buffer.nil?
     end
 
     def capture_later(&block)
       engine = current_engine
-      proc { |*a| with_engine(engine) { @capture = capture(*a, &block) }}
```

Comment:
Why did you remove `@capture` from this point? We have a possibility of the user who depends on this behavior.

Reply:
I've put @capture back in the off chance someone has made use of this variable.